src/rules/noSelfAssignRule.ts

```diff
@@ -0,0 +1,210 @@
+            }
+
+        }
+
+        if (isObjectLiteralExpression(left) && isObjectLiteralExpression(right)) {
+
+            /*
+            * Gets the index of the last spread property.
+            * It's possible to overwrite properties followed by it.
+            */
+            let startRight = 0;
+
+            for (let i = right.properties.length - 1; i >= 0; i--) {
+                if (isSpreadAssignment(right.properties[i])) {
+                    startRight = i + 1;
+                    break;
+                }
+            }
+
+            for (const leftProp of left.properties) {
+                for (let j = startRight; j < right.properties.length; j++) {
+                    doCheck(leftProp, right.properties[j]);
+                }
+            }
+        }
+
+        // ({a} = {a});
+        if (isShorthandPropertyAssignment(left) && left.objectAssignmentInitializer == undefined
+            && isShorthandPropertyAssignment(right)) {
+            doCheck(left.name, right.name);
+        }
+
+        // a.b = a.b
+        if (ctx.options.props && isPropertyAccessExpression(left) && isPropertyAccessExpression(right) && isSameMember(left, right)) {
+            report(right);
+            return;
+        }
+
+        // a["b"] = a["b"]
+        if (ctx.options.props && isElementAccessExpression(left) && isElementAccessExpression(right)
+            && isSameElementAccess(left, right)) {
+            report(right);
+        }
+    }
+}
+
+function isSameElementAccess(left: ts.ElementAccessExpression, right: ts.ElementAccessExpression) {
+    const leftExp = left.expression;
+    const rightExp = right.expression;
+    if (isSameIdentifier(leftExp, rightExp)) {
+        const leftArg = left.argumentExpression;
+        const rightArg = right.argumentExpression;
+
+        if (leftArg != undefined && rightArg != undefined) {
+            if (isIdentifier(leftArg) && isIdentifier(rightArg)) {
+                return leftArg.text === rightArg.text;
+            }
+
+            if (isLiteralExpression(leftArg) && isLiteralExpression(rightArg)) {
+                return leftArg.text === rightArg.text;
+            }
+        }
+    }
+
+    return false;
+}
+
+function isSameIdentifier(left: ts.Node, right: ts.Node) {
+    return isIdentifier(left) && isIdentifier(right) && left.text === right.text;
+}
+
+function isSameMember(left: ts.PropertyAccessExpression, right: ts.PropertyAccessExpression): boolean {
+    if (!isSameProperty(left, right)) {
+        return false;
+    }
+
+    const leftExp = left.expression;
+    const rightExp = right.expression;
+
+    if (isPropertyAccessExpression(leftExp) && isPropertyAccessExpression(rightExp)) {
+        return isSameMember(leftExp, rightExp);
+    }
+
+    return isIdentifier(leftExp) && isIdentifier(rightExp) && leftExp.text === rightExp.text;
+}
+
+function isSameProperty(left: ts.Node, right: ts.Node) {
+    if (isPropertyAccessExpression(left) && isPropertyAccessExpression(right)
```

Comment:
Nit: just directly return the result of this expression. `return isPropertyAccessExpression(left) && ...`